primitives/src/calendar.rs

```diff
@@ -0,0 +1,321 @@
+            }
+            CalendarUnit::Year => FixedOrVariableCalendarUnit::Variable(VariableCalendarUnit::Year),
+        }
+    }
+}
+
+/// The schedule of an asset checkpoint containing the start time `start` and the optional period
+/// `period` in case the checkpoint is to recur after `start` at regular intervals.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Clone, Copy, Debug, Default, PartialEq, Eq)]
+pub struct CheckpointSchedule {
+    /// Unix time in seconds (UTC).
+    pub start: u64,
+    /// The period at which the checkpoint is set to recur after `start`.
+    pub period: CalendarPeriod,
+}
+
+impl CheckpointSchedule {
+    /// Computes the next checkpoint for the schedule given the current timestamp in seconds UTC.
+    pub fn next_checkpoint(&self, now_as_secs_utc: u64) -> Option<u64> {
+        if self.start > now_as_secs_utc {
+            // The start time is in the future.
+            return Some(self.start);
+        }
+        // The start time is in the past.
+        let multiplier = self.period.multiplier;
+        if multiplier == 0 {
+            // The period is empty while the start time has already passed.
+            return None;
+        }
+        match self.period.as_fixed_or_variable() {
+            FixedOrVariableCalendarUnit::Fixed(period_as_secs) => {
+                // The period is of fixed length in seconds UTC.
+                let secs_since_start = now_as_secs_utc - self.start;
+                let elapsed_periods: u64 = secs_since_start / period_as_secs;
+                Some(secs_since_start + period_as_secs * (elapsed_periods + 1))
+            }
+            FixedOrVariableCalendarUnit::Variable(variable_unit) => {
+                // The period is of variable length.
+                let date_time_start =
+                    NaiveDateTime::from_timestamp(i64::try_from(self.start).ok()?, 0);
+                let date_start = date_time_start.date();
+                let year_start = date_start.year();
+                let month_start = date_start.month();
+                let day_start = date_start.day();
+                let date_time_now =
+                    NaiveDateTime::from_timestamp(i64::try_from(now_as_secs_utc).ok()?, 0);
+                let date_now = date_time_now.date();
+                match variable_unit {
+                    VariableCalendarUnit::Month => {
+                        // Convert the multiplier to match the type of month.
+                        let multiplier = u32::try_from(multiplier).ok()?;
+                        let year_diff = u32::try_from(date_now.year() - year_start).ok()?;
+                        // Convert months to base 12.
+                        let month_start_base12 = month_start - 1;
+                        let month_now_base12 = date_now.month() - 1;
+                        let elapsed_months = year_diff * 12 + month_now_base12 - month_start_base12;
```

Comment:
Is this comment accurate? You are indeed computing the number of elapsed months (although `- 1` is on both sides so it has no impact), but this is still the number of months in base10, just fixing the issue that chrono's `.month()` is 1-indexed rather than `0-indexed`. Maybe adjust the comment to discuss `.month()`'s behavior instead? Even better, you can use `.month0()` to get it 0-indexed.